ryzen_smu: init at 0.1.5, ryzen_monitor_ng: init at 2.0.5

PhDyellow opened the pull request:
## Description of changes

ryzen_monitor_ng is a tool for monitoring power information on desktop AMD processors.
It can also set parameters for the CPU, so can offer some of the features of Ryzen Master on Linux.

It depends on the linux kernel driver ryzen_smu.

ryzen_smu includes a userspace program `monitor_cpu`.

https://gitlab.com/mann1x/ryzen_smu

https://github.com/mann1x/ryzen_monitor_ng


<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [X] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
maintainers: add phdyellow

Commit:
doc: add information about ryzen-monitor-ng and ryzen-smu modules

PhDyellow commented:
Tested on a 5900X CPU in a clevo laptop.

Cryolitia reviewed (commented):
I would like to introduce these two patches to support Phonix and Rembrandt.

https://github.com/Cryolitia/nixpkgs/commit/9aecdf4daaedca78bb47cea2442f5c5c96176b30#diff-2cc356432ce45563cde8df0d10e36f41c3524284e9ceb526c281f3a61219ffdaR44

Would you like to co-maintain the module ryzen-smu, together. I take the liberty of making this request. I would be honored if you would accept it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ryzenadj-and-ryzen-smu/31326/4

PhDyellow commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,42 @@
+  version = "0.1.5";
+
+  src = fetchFromGitLab {
+    owner = "mann1x";
```
The ryzen_smu programs have gone through a lot of forks themselves.

[ryzen_smu](https://gitlab.com/leogx9r/ryzen_smu) is the original ryzen_smu. It included `monitor_cpu`. [ryzen_monitor](https://github.com/hattedsquirrel/ryzen_monitor) extended `monitor_cpu` and [ryzen_monitor_ng](https://github.com/mann1x/ryzen_monitor_ng) extended `ryzen_monitor`.

To support [ryzen_monitor_ng](https://github.com/mann1x/ryzen_monitor_ng), `mann1x` forked `ryzen_smu` but hasn't merged the changes. See [here](https://github.com/mann1x/ryzen_monitor_ng#building) and [here](https://gitlab.com/leogx9r/ryzen_smu/-/issues/26#note_1666173169).

However, when I built `ryzen_monitor_ng` from [mann1x's repo](https://github.com/mann1x/ryzen_monitor_ng) it wouldn't run because of a version mismatch in `libsmu`. [kvic-z's fork of ryzen_monitor_ng](https://github.com/kvic-z/ryzen_monitor_ng) corrects the version mismatch and adds Matisse.

The patches you suggest come from yet another fork.

I do like your approach of patching the original repos. I hope it doesn't lead to conflicts between `mann1x's` fork and `moson-mo's`. Worth a try, even if it leads to a lot of patches.

Cryolitia replied:
It's so complex that I have just realized. I have noticed that the origin repo hasn't updated since half year ago. Do you think if it's a good idea that we maintain a fork in person, picking PRs and testing, then releasing.

Anyway, please at least leave a comment in the Nix packaging source code about it as long as you choose a repo which is not the origin one. Otherwise, other people who don't see this discussion would get confused.

PhDyellow replied:
It is looking like we will need to maintain a repo with all the patches. PRs are not getting merged upstream quickly. Are you willing to set up the repo? I will change the `src` if you do.

Cryolitia replied:
> To support [ryzen_monitor_ng](https://github.com/mann1x/ryzen_monitor_ng), `mann1x` forked `ryzen_smu` but hasn't merged the changes. See [here](https://github.com/mann1x/ryzen_monitor_ng#building) and [here](https://gitlab.com/leogx9r/ryzen_smu/-/issues/26#note_1666173169).

I have noticed the same commits with different sha256 in mann1x's repo and leogx9r's repo, such as https://gitlab.com/leogx9r/ryzen_smu/-/commit/e61177d0ddaebfaeca52094b20a2289287a0838b and https://gitlab.com/mann1x/ryzen_smu/-/commit/adaf53c77dc68b99643ab7af729eb9c592cf5e01. Has it already rebased or I  miss something other place.

PhDyellow commented on nixos/modules/hardware/ryzen-smu.nix:
```diff
@@ -0,0 +1,26 @@
+
+let
+  cfg = config.hardware.ryzen_smu;
+  ryzen_smu = config.boot.kernelPackages.ryzen_smu;
```
Does `ryzenadj` use `ryzen_smu` at all?
Does `ryzen_smu` work with laptop CPUs?
 I put this PR together because `RyzenAdj` refused to work with my desktop AMD CPU.

Cryolitia replied:
> I would like to add a warn for users installing `ryzenadj`

What's your point of view @rhendric ? I'm sorry to bother you, I thought you might be interested in this PR as `ryzenadj` 's maintainer.

rhendric replied:
I don't have a great picture of all the CPUs that are and are not supported by all the different tweaking utilities. I use RyzenAdj on a 5700U and I don't seem to need any modules to enable its full capabilities there (I don't have any `nan`s in my `ryzenadj -i` output), so this warning wouldn't seem to be applicable to everyone.

PhDyellow replied:
I know RyzenAdj only supports laptop CPUs (with a "U"?), so desktop CPU users can't use it. For example, my 5900x is not supported by RyzenAdj, as it is a desktop chip.

I am fairly sure RyzenAdj does not use `ryzen_smu` at all, [there is no reference to libsmu in the repo](https://github.com/search?q=repo%3AFlyGoat%2FRyzenAdj%20libsmu&type=code). RyzenAdj has it's own smu library: `ryzen_nb_smu`.

Rather than warning users when they happen to have one package installed, we probably need a wiki to recommend packages users can try. [The arch wiki is actually pretty good](https://wiki.archlinux.org/title/Ryzen), it mentions both RyzenAdj and ryzen_smu, and led me to a number of other ryzen utilities.

Cryolitia replied:
> Rather than warning users when they happen to have one package installed, we probably need a wiki to recommend packages users can try. [The arch wiki is actually pretty good](https://wiki.archlinux.org/title/Ryzen), it mentions both RyzenAdj and ryzen_smu, and led me to a number of other ryzen utilities.

I agree  with it now.

PhDyellow commented on pkgs/os-specific/linux/ryzen-smu/monitor-cpu.nix:
```diff
@@ -0,0 +1,36 @@
+{ lib, stdenv, fetchFromGitLab }:
```
Good point. May I modify this PR and model it after your patches? I wasn't sure having `monitor-cpu` was even a good idea at all, it gives incorrect readings on my system.

Cryolitia replied:
I‘m not sure. Seems that `ryzen_smu` supports my 7840U, but `monitor-cpu` doesn't. I failed to test it, and will manager to get another AMD device this weekend.

PhDyellow commented:
I would be happy to co-maintain it with you. Given the number of forks and patches that ryzen_smu uses, I'll need the help.

Your patches contain some better nix code than I have written, but it looks like you wrote it independently to this PR. May I pull in the relevant parts of your patches and merge/squash/rebase the whole PR?

Cryolitia commented:
> I would be happy to co-maintain it with you. Given the number of forks and patches that ryzen_smu uses, I'll need the help.
> 
> Your patches contain some better nix code than I have written, but it looks like you wrote it independently to this PR. May I pull in the relevant parts of your patches and merge/squash/rebase the whole PR?

Feel free to use my source code ^_^

Cryolitia commented:
Tested on multi AMD processors, results here: https://gitlab.com/leogx9r/ryzen_smu/-/merge_requests/12#note_1704774029

TLDR: Each processor has different issues, and no one processor has all the features working properly.

PhDyellow commented:
Thanks @Cryolitia, I have updated this PR with your recommendations. The patches apply cleanly to the `mann1x/ryzen_smu` fork, so I think we can get away without starting yet another repo.

If it looks good to you, I will squash the commits down for a clean merge.

PhDyellow commented:
Please test `ryzen_smu` with `ryzen_monitor_ng`, it produces more accurate readings than `monitor_cpu`

Cryolitia commented:
> It's so complex that I have just realized. I have noticed that the origin repo hasn't updated since half year ago. Do you think if it's a good idea that we maintain a fork in person, picking PRs and testing, then releasing.

Just because that I have found other interesting PRs , such as https://gitlab.com/leogx9r/ryzen_smu/-/merge_requests/13

Cryolitia commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,29 @@
+let
+  cfg = config.hardware.cpu.amd.ryzen_smu;
+  ryzen_smu = config.boot.kernelPackages.ryzen_smu;
+  ryzenadjCheck = lib.warnIf (builtins.elem pkgs.ryzenadj config.environment.systemPackages)
+    "ryzenadj is in environment.systemPackages, try hardware.cpu.amd.ryzen-smu and programs.ryzen_monitor_ng for more features.";
```
This doesn't seem to work at all, it seems either I don't understand Nix thoroughly, or this just doesn't make sense. This is my fault please forgive me.

PhDyellow replied:
My guess is that it isn't getting used in the function, so lazy evaluation means it never runs. 

I am going to take it out though, because a wiki listing both ryzenadj and ryzen_smu would be more useful.

Cryolitia commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,42 @@
+  };
+
+  buildPhase = ''
+    runHook preBuild
+    make clean
+    rm src/ryzen_monitor
+    make
+    runHook postBuild
```
```suggestion
    runHook preBuild

    make clean
    rm src/ryzen_monitor
    make

    runHook postBuild
```

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,91 @@
+    hash = "sha256-OQyPqNL8Cd7XQfY8YiOVYqOG7RIu5l+wJoK2FDBg5eg=";
+  };
+
+  monitor-cpu = stdenv.mkDerivation (finalAttrs: {
```
don' forget to patch this.

https://gitlab.com/leogx9r/ryzen_smu/-/merge_requests/12#note_1700621578

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,91 @@
+  '';
+
+  meta = with lib; {
+    broken = stdenv.isAarch64;
```
Have you tested it to building (just building) on aarch64?

Cryolitia replied:
just it's so confused to mark it as broken instead of unsupported platform

PhDyellow replied:
I get it now, "broken" means it should work, but does not.

I will set the platform to x86_64 instead, so aarch64 is unsupported.

Cryolitia commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,42 @@
+    homepage = "https://github.com/mann1x/ryzen_monitor_ng";
+    license = licenses.agpl3Only;
+    platforms = [ "x86_64-linux" ];
+    maintainers = [ lib.maintainers.phdyellow ];
```
please add a `mainProgram`

Cryolitia commented:
tested 
![image](https://github.com/NixOS/nixpkgs/assets/23723294/3e7f5187-91e0-4eaa-93a2-b73e98b70196)

PhDyellow commented:
Thanks for testing. It looks like the patches aren't working yet.

Can you share the output of:

```
sudo cat /sys/kernel/ryzen_smu_drv/codename
```

PhDyellow commented:
Ah, got it! `monitor_cpu` wasn't patched, so it wouldn't accept Phoenix CPUs.

PhDyellow commented:
`ryzen_monitor_ng` has even more forks with useful patches: https://github.com/plasmin/ryzen_monitor_ng

`ryzen_monitor_ng` needs the PM table to be explicitly defined for each processor. You might be able to get it to work with 


```
sudo ryzen_monitor -f 0x380805
```

This will force it to use the 5900x PM table, which is almost certainly wrong : )

Adding the PM table for Phoenix will require a PR to `ryzen_monitor_ng`.

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,95 @@
+  '';
+
+  meta = with lib; {
+    #broken = stdenv.isAarch64;
+    description = "A Linux kernel driver that exposes access to the SMU (System Management Unit) for certain AMD Ryzen Processors";
+    homepage = "https://gitlab.com/leogx9r/ryzen_smu";
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ Cryolitia phdyellow ];
+    platforms = platforms.linux;
+    mainProgram = "monitor_cpu";
```
```suggestion
    description = "A Linux kernel driver that exposes access to the SMU (System Management Unit) for certain AMD Ryzen Processors";
    homepage = "https://gitlab.com/leogx9r/ryzen_smu";
    license = licenses.gpl2Plus;
    maintainers = with maintainers; [ Cryolitia phdyellow ];
    platforms = [ "x86_64-linux" ];
    mainProgram = "monitor_cpu";
```

> The difference between a package being unsupported on some system and being broken is admittedly a bit fuzzy. If a program ought to work on a certain platform, but doesn’t, the platform should be included in meta.platforms, but marked as broken with e.g. meta.broken = !hostPlatform.isWindows. Of course, this begs the question of what “ought” means exactly. That is left to the package maintainer.

see https://nixos.org/manual/nixpkgs/stable/#sec-allow-unsupported-system

cguentherTUChemnitz commented:
Please bring this merge through. Undervolting capabilities for the 7840HS in a notebook is quite the last thing for nixos being the optimal setup for the notebook

Cryolitia commented:
> Please bring this merge through. Undervolting capabilities for the 7840HS in a notebook is quite the last thing for nixos being the optimal setup for the notebook.
> 
> update: Just for clarification: as far as i understand, the ryzen_smu module is the missing piece to get ryzenadj working to undervolt the chip, right? At least, when i run `sudo ryzenadj -i` i get the following printout:
> 
> ```
> CPU Family: Phoenix
> SMU BIOS Interface Version: 12
> Version: v0.14.0 
> failed to get /sys/kernel/ryzen_smu_drv/pm_table: No such file or directory
> failed to map /dev/mem: Operation not permitted
> If you don't want to change your memory access policy, you need a kernel module for this task.
> We do support usage of this kernel module https://gitlab.com/leogx9r/ryzen_smu
> Unable to get memory access
> Unable to init power metric table: -5, this does not affect adjustments because it is only needed for monitoring.
> ```

Sorry for the holiday delay and my

Cryolitia commented:
> Please bring this merge through. Undervolting capabilities for the 7840HS in a notebook is quite the last thing for nixos being the optimal setup for the notebook.

So sorry for the holiday and my busy final month of college resulting in my AFK. I would manage to continu working on this soon.

You could try my NUR module https://github.com/Cryolitia/nur-packages/blob/master/modules/ryzen-smu.nix , it's one of the bases of this PR. Though it contains many defects but I hope it's finally at least basically usable.

PhDyellow commented:
> Please bring this merge through. Undervolting capabilities for the 7840HS in a notebook is quite the last thing for nixos being the optimal setup for the notebook.
> 
> update: Just for clarification: as far as i understand, the ryzen_smu module is the missing piece to get ryzenadj working to undervolt the chip, right? At least, when i run `sudo ryzenadj -i` i get the following printout:
> 
> ```
> CPU Family: Phoenix
> SMU BIOS Interface Version: 12
> Version: v0.14.0 
> failed to get /sys/kernel/ryzen_smu_drv/pm_table: No such file or directory
> failed to map /dev/mem: Operation not permitted
> If you don't want to change your memory access policy, you need a kernel module for this task.
> We do support usage of this kernel module https://gitlab.com/leogx9r/ryzen_smu
> Unable to get memory access
> Unable to init power metric table: -5, this does not affect adjustments because it is only needed for monitoring.
> ```

Thanks @cguentherTUChemnitz, I can't use RyzenAdj and I missed that comment when I glanced at the RyzenAdj codebase.

It does look like RyzenAdj can use ryzen_smu. 

Once this PR is merged, it would be worth opening an issue to ask the RyzenAdj maintainer to enable ryzen_smu.

PhDyellow commented:
As @Cryolitia said, you don't have to wait for this PR to merge before using this fork. Use the NUR module, or set nixpkgs to one of my forks in your config, this one has been rebased to more recent master: https://github.com/PhDyellow/nixpkgs/tree/ryzen_smu_rebased. 

I often set nixpkgs to my fork until a PR merges.

Cryolitia commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,45 @@
+
+  meta = with lib; {
+    description = "Access Ryzen SMU information exposed by the ryzen_smu driver";
+    homepage = "https://github.com/mann1x/ryzen_monitor_ng";
```
```suggestion
    homepage = "https://github.com/leogx9r/ryzen_monitor_ng";
```

PhDyellow replied:
leogx9r is not an author of `ryzen_monitor_ng`. It's a very confusing network of repos : )

Cryolitia replied:
> leogx9r is not an author of `ryzen_monitor_ng`. It's a very confusing network of repos : )

so sorry……my fault

Cryolitia commented:
excuse but is it ready for review now(?

PhDyellow commented:
Sorry, I was waiting for it to build on my machine. 

Yes, it is ready for you to review now. If it looks good to you, I will squash the commits down and then this PR will be ready.

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,80 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+
+    description = "A Linux kernel driver that exposes access to the SMU (System Management Unit) for certain AMD Ryzen Processors";
+
+    homepage = "https://gitlab.com/leogx9r/ryzen_smu";
+
+    license = licenses.gpl2Plus;
+
+    maintainers = with maintainers; [ Cryolitia phdyellow ];
+
+    platforms = [ "x86_64-linux" ];
+
+    mainProgram = "monitor_cpu";
```
```suggestion
  meta = with lib; {
    description = "A Linux kernel driver that exposes access to the SMU (System Management Unit) for certain AMD Ryzen Processors";
    homepage = "https://gitlab.com/leogx9r/ryzen_smu";
    license = licenses.gpl2Plus;
    maintainers = with maintainers; [ Cryolitia phdyellow ];
    platforms = [ "x86_64-linux" ];
    mainProgram = "monitor_cpu";
```

PhDyellow replied:
Agree

Cryolitia commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,45 @@
+{ lib, stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation rec {
+  pname = "ryzen_monitor_ng";
```
```suggestion
  pname = "ryzen-monitor-ng";
```

https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md?plain=1#L357

MinerSebas replied:
Your proposed change conflicts with this rule:

https://github.com/NixOS/nixpkgs/blob/535caea95c5957391e878fe43fdaf81e0946e869/pkgs/README.md?plain=1#L347

And the name of the upstream package uses `_`

Cryolitia replied:
My apology

PhDyellow replied:
I will leave it as is.

Cryolitia commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,45 @@
+
+stdenv.mkDerivation rec {
+  pname = "ryzen_monitor_ng";
+  version = "2.0.5";
```
```suggestion
  version = "unstable-2023-11-05";
```

Cryolitia replied:
plasmin’s repo doesn't have any tag, it makes it difficult to find out the latest versio. I don't know if it should have a version from a more original repo, but don't oppose it.

PhDyellow replied:
I guess I am using the fork as a small set of patches.

I can revert to using the orginal repo, and pull in the patches from the forks if that seems better to you.

PhDyellow replied:
Then the version would be from the original repo, but it would still be unstable.

Cryolitia commented on nixos/modules/programs/ryzen-monitor-ng.nix:
```diff
@@ -0,0 +1,39 @@
+
+  };
+
+  config = mkIf cfg.enable (mkMerge [
```
why `mkMerge` here?

PhDyellow replied:
Copy paste from another package.

That's a terrible reason : P

I'm going to read up on it now.

PhDyellow replied:
https://nixos.org/manual/nixos/stable/#sec-option-definitions-merging

Not needed, removing.

Thanks!

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,80 @@
+}:
+
+let
+  smu_version = "0.1.5";
```
```suggestion
  smu_version = "unstable-2024-01-03";
```

PhDyellow replied:
Agree

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,80 @@
+  };
+
+  monitor-cpu = stdenv.mkDerivation (finalAttrs: {
+    pname = "monitor_cpu";
```
```suggestion
    pname = "monitor-cpu";
```

Cryolitia commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,80 @@
+
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "ryzen_smu-${kernel.version}";
```
```suggestion
  pname = "ryzen-smu-${kernel.version}";
```

Cryolitia commented on pkgs/top-level/all-packages.nix:
```diff
@@ -41965,6 +41965,8 @@ with pkgs;
 
   ryzenadj = callPackage ../os-specific/linux/ryzenadj { };
 
+  ryzen_monitor_ng = callPackage ../os-specific/linux/ryzen-monitor-ng { };
```
```suggestion
  ryzen-monitor-ng = callPackage ../os-specific/linux/ryzen-monitor-ng { };
```

Cryolitia commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -543,6 +543,8 @@ in {
 
     ithc = callPackage ../os-specific/linux/ithc { };
 
+    ryzen_smu = callPackage ../os-specific/linux/ryzen-smu { };
```
```suggestion
    ryzen-smu = callPackage ../os-specific/linux/ryzen-smu { };
```

Cryolitia commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,26 @@
+
+  config = mkIf cfg.enable {
+    boot.kernelModules = [ "ryzen_smu" ];
+    boot.extraModulePackages = [ ryzen_smu ];
+    environment.systemPackages = [ ryzen_smu ];
```
```suggestion
    boot.extraModulePackages = [ ryzen-smu ];
    environment.systemPackages = [ ryzen-smu ];
```

MinerSebas reviewed (changes requested):
For completeness, I think there should also be `options.hardware.cpu.amd.ryzen_smu.package` and `programs.ryzen_monitor_ng.package` Options to change the used Package.

Edit: Thinking again i don't know how common it is to override a kernel module using an Option, so a `options.hardware.cpu.amd.ryzen_smu.package` Option could perhaps be skipped.

Cryolitia commented:
> For completeness, I think there should also be `options.hardware.cpu.amd.ryzen_smu.package` and `programs.ryzen_monitor_ng.package` Options to change the used Package.

I agree with that.

> Edit: Thinking again i don't know how common it is to override a kernel module using an Option, so a `options.hardware.cpu.amd.ryzen_smu.package` Option could perhaps be skipped.

It can be necessary, specifically for this package, the upstream repository has not been maintained for more than half a year, and there are many different forks scattered everywhere. Adding `package` option may help maintainers and others to select source code from different repositories and test its functionality.

PhDyellow commented on pkgs/os-specific/linux/ryzen-monitor-ng/default.nix:
```diff
@@ -0,0 +1,45 @@
+    homepage = "https://github.com/mann1x/ryzen_monitor_ng";
+    license = licenses.agpl3Only;
+    platforms = [ "x86_64-linux" ];
+    maintainers = [ lib.maintainers.phdyellow ];
```
Agree

PhDyellow commented:
I will put in package options for both. I don't know what someone would need to do to include a kernel module from a flake or similar, but I'm sure it is possible.

PhDyellow commented:
Something is breaking.

PhDyellow commented:
https://nixos.wiki/wiki/Linux_Kernel

As a kernel package,  `ryzen_smu` actually has many packages in nixpkgs, one for each kernel. The "default" package is the one associated with your kernel. I can't see a way to automatically select the `ryzen_smu` package that matches your kernel, or a good way to explain to users how to make sure their package is using the correct kernel.

I did try 
```
    package = mkPackageOption pkgs.boot.kernelPackages "ryzen_smu" {
      pkgsText = "pkgs.boot.kernelPackages";
    };
```

but got an error that `kernelPackages` didn't exist. I also tried 

```
    package = mkPackageOption pkgs.linuxPackages "ryzen_smu" {
      pkgsText = "pkgs.linuxPackages";
    };
```

I thought that might work, because `linuxPackages` is more clearly a package set than `boot.kernelPackages`. The error now was that it targeted the wrong kernel version. For my machine, I could get it to work with `linuxPackages_latest` but that is clearly not a general solution.

I think the best way forward is that anyone wanting to use a different `ryzen_smu` package should be directed to the nixos wiki, which explains how to replace kernel modules with out-of-tree packages. 

Open to other solutions, but for now I don't have time to force this to work.

PhDyellow commented:
A reference to another kernel module that has a package option would be enough, I could follow it as an example from there.

Cryolitia reviewed (approved):
##### Nixpkgs reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on ARCHITECTURE
- [x] executables tested on ARCHITECTURE (partly, some excutables not compatible with my machine)
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

##### NixOS reviewed points

- [x] module path fits the guidelines
- [x] module tests succeed on ARCHITECTURE
- [x] options have appropriate types
- [x] options have default
- [x] options have example
- [x] options have descriptions
- [x] No unneeded package is added to `environment.systemPackages`
- [x] `meta.maintainers` is set
- [x] module documentation is declared in `meta.doc`

PhDyellow commented:
Ok, I will squash the commits down

PhDyellow commented:
Sorry @Cryolitia, I think github wants you to re-approve the PR after I squashed the commits.

Otherwise I'm not sure who needs to merge the PR.

Cryolitia commented:
> Sorry @Cryolitia, I think github wants you to re-approve the PR after I squashed the commits.
> 
> Otherwise I'm not sure who needs to merge the PR.

Just be patient, waiting for a committee or more other people to review. The label is enough.

nu-nu-ko commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,26 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
```suggestion
```
to address #208242 for this module, uses of `with` such as this should be avoided.
consider `inherits` where you find yourself using `lib.foo` excessively.
[this](https://nix.dev/guides/best-practices.html#with-scopes)  nix.dev page on best practices with `with` may be helpful

PhDyellow replied:
Done

nu-nu-ko commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,26 @@
+in
+{
+  options.hardware.cpu.amd.ryzen_smu = {
+    enable = mkEnableOption (lib.mdDoc ''
```
```suggestion
    enable = mkEnableOption ''
```
`lib.mdDoc` is now just an alias and can be safely removed everywhere.
see https://github.com/NixOS/nixpkgs/commit/d36f950d405369abca5144b14a201bc25c4a080a and https://github.com/NixOS/nixpkgs/pull/237557

nu-nu-ko commented on nixos/modules/programs/ryzen-monitor-ng.nix:
```diff
@@ -0,0 +1,40 @@
+{ pkgs, config, lib, ... }:
+
+with lib;
```
this use of with should also be removed.

PhDyellow replied:
Done

PhDyellow commented:
Rebased to master to comply with `pkgs/by-name` file structure.

wegank commented on pkgs/by-name/ry/ryzen_monitor_ng/package.nix:
```diff
@@ -0,0 +1,45 @@
+{ lib, stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation rec {
+  pname = "ryzen_monitor_ng";
```
`ryzen-monitor-ng` is preferred over `ryzen_monitor_ng`, see [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming).

PhDyellow replied:
My understanding of `pname` is that it should match upstream:

https://github.com/NixOS/nixpkgs/blob/57e6b3a9e4ebec5aa121188301f04a6b8c354c9b/pkgs/README.md?plain=1#L349

For all file names and paths, dashes should be used:

https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#code-conventions


Upstream uses `ryzen_monitor_ng`, so I have set `pname=ryzen_monitor_ng` and used `ryzen-monitor-ng` for all file and path names, and for the entries in `all-packages.nix`.

PhDyellow replied:
However, I see I have used underscores in the nixos modules and `linux_kernels.nix`. Do you recommend changing those?

SuperSandro2000 replied:
> However, I see I have used underscores in the nixos modules and `linux_kernels.nix`. Do you recommend changing those?

Changing existing names is always harder than new ones.

> My understanding of `pname` is that it should match upstream:

That mainly focuses on the name. We should not name this package ``amd-monitoring-new``

PhDyellow replied:
Thanks for clarifying, I'll change underscores to hyphens.

wegank commented on pkgs/by-name/ry/ryzen_monitor_ng/package.nix:
```diff
@@ -0,0 +1,45 @@
+{ lib, stdenv, fetchFromGitHub }:
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation {
```

wegank commented on pkgs/by-name/ry/ryzen_monitor_ng/package.nix:
```diff
@@ -0,0 +1,45 @@
+    owner = "plasmin";
+    repo = "ryzen_monitor_ng";
+    rev = "8b7854791d78de731a45ce7d30dd17983228b7b1";
+    sha256 = "sha256-fcW2fEsCFliRnMFnboR0jchzVIlCYbr2AE6AS06cb6o=";
```
Probably also `hash` here, like what is put in `ryzen-smu`.

PhDyellow replied:
Agreed

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -49,6 +49,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 - [pyLoad](https://pyload.net/), a FOSS download manager written in Python. Available as [services.pyload](#opt-services.pyload.enable)
 
 - [maubot](https://github.com/maubot/maubot), a plugin-based Matrix bot framework. Available as [services.maubot](#opt-services.maubot.enable).
```
```suggestion
- [maubot](https://github.com/maubot/maubot), a plugin-based Matrix bot framework. Available as [services.maubot](#opt-services.maubot.enable).

```

SuperSandro2000 commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,28 @@
+{
+  options.hardware.cpu.amd.ryzen_smu = {
+    enable = mkEnableOption ''
+        A linux kernel driver that exposes access to the SMU (System Management Unit) for certain AMD Ryzen Processors.
```
mkEnableOption will prefix ``Wether to`` and suffix ``.``.

SuperSandro2000 commented on nixos/modules/hardware/cpu/amd-ryzen-smu.nix:
```diff
@@ -0,0 +1,28 @@
+{ config
+, lib
+, pkgs
```
```suggestion
```
unused

SuperSandro2000 commented on nixos/modules/programs/ryzen-monitor-ng.nix:
```diff
@@ -0,0 +1,44 @@
+
+  };
+
+  config = mkIf cfg.enable {
+      environment.systemPackages = [ cfg.package ];
+      hardware.cpu.amd.ryzen_smu.enable = true;
+  };
+
+  meta.maintainers = with lib.maintainers; [ Cryolitia phdyellow ];
+
+}
```
```suggestion
  config = mkIf cfg.enable {
    environment.systemPackages = [ cfg.package ];
    hardware.cpu.amd.ryzen_smu.enable = true;
  };

  meta.maintainers = with lib.maintainers; [ Cryolitia phdyellow ];
}
```
we use 2 spaces to intend in nix files.

SuperSandro2000 commented on nixos/modules/programs/ryzen-monitor-ng.nix:
```diff
@@ -0,0 +1,44 @@
+  options = {
+
+    programs.ryzen_monitor_ng = {
+      enable =  mkEnableOption ''
```
This has the same prefixing problem as the other option.

SuperSandro2000 commented on nixos/modules/programs/ryzen-monitor-ng.nix:
```diff
@@ -0,0 +1,44 @@
+, ...
+}:
+
+let
+
+  inherit (lib) mkEnableOption mkPackageOption mkIf;
+
+  cfg = config.programs.ryzen_monitor_ng;
+
+in
+{
+
+  options = {
+
```
```suggestion
let
  inherit (lib) mkEnableOption mkPackageOption mkIf;
  cfg = config.programs.ryzen_monitor_ng;
in
{
  options = {
```
nit

SuperSandro2000 commented on pkgs/by-name/ry/ryzen_monitor_ng/package.nix:
```diff
@@ -0,0 +1,45 @@
+  buildPhase = ''
+    runHook preBuild
+
+    make clean
+    rm src/ryzen_monitor
+    make
```
Can't we use makeTarget and preBuild?

PhDyellow replied:
Thanks for the pointer, code should be clearer now.

SuperSandro2000 commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,74 @@
+}:
+
+let
+  smu_version = "0.1.5-unstable-2024-01-03";
+
+  ## Upstream has not been merging PRs.
+  ## Nixpkgs maintainers are providing a
+  ## repo with PRs merged until upstream is
+  ## updated.
+  smu_src = fetchFromGitHub {
```
```suggestion
  version = "0.1.5-unstable-2024-01-03";

  ## Upstream has not been merging PRs.
  ## Nixpkgs maintainers are providing a
  ## repo with PRs merged until upstream is
  ## updated.
  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+  monitor-cpu = stdenv.mkDerivation (finalAttrs: {
+    pname = "monitor_cpu";
+    version = smu_version;
+
+    src = smu_src;
```
```suggestion
    inherit version src;
```

SuperSandro2000 commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,74 @@
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "ryzen_smu-${kernel.version}";
+  version = smu_version;
+
+  src = smu_src;
```
```suggestion
    inherit version src;
```

SuperSandro2000 commented on pkgs/os-specific/linux/ryzen-smu/default.nix:
```diff
@@ -0,0 +1,74 @@
+  });
+
+in
+stdenv.mkDerivation (finalAttrs: {
```
```suggestion
stdenv.mkDerivation {
```
that's not used

Aleksanaa reviewed (approved):
Lgtm. Please squash your commits.

Aleksanaa commented:
Sorry, the nixos module should be added in another commit (not package), and follow commit conventions in https://github.com/NixOS/nixpkgs/tree/master/nixos#commit-conventions

PhDyellow commented:
> Why this suggested change was reverted? [#271342 (comment)](https://github.com/NixOS/nixpkgs/pull/271342#discussion_r1536915220)

Thanks for catching that, I didn't mean to revert that.

> Sorry, the nixos module should be added in another commit (not package), and follow commit conventions in https://github.com/NixOS/nixpkgs/tree/master/nixos#commit-conventions

Makes sense, I will split up the commits as soon as I can.

Commit:
ryzen-smu: init at 0.1.5

A Linux kernel driver that exposes access to the
SMU (System Management Unit) for certain AMD Ryzen Processors.

Contains monitor_cpu, a userspace tool for viewing info.

Using fork of original to match ryzen_monitor_ng, a more advanced
userspace tool for accessing the SMU via this kernel module,
planned for a later commit.

Commit:
nixos/ryzen-smu: init module

Provide a module for installing ryzen_smu, a Linux kernel driver
that exposes access to the SMU (System Management Unit) for
certain AMD Ryzen Processors.

Installs monitor_cpu, a userspace tool for viewing info.

Using fork of original to match ryzen_monitor_ng, a more advanced
userspace tool for accessing the SMU via this kernel module,
planned for a later commit.

Commit:
ryzen-monitor-ng: init at 2.0.5

A userspace tool for setting and getting AMD CPU power
and performance parameters.

Relies on the ryzen_smu kernel module.

Commit:
nixos/ryzen-monitor-ng: init module

A userspace tool for setting and getting AMD CPU power
and performance parameters.

The module adds `ryzen_monitor_ng` to `environment.systemPackages` and
enables the `ryzen-smu` module, as `ryzen_monitor_ng` requires the
`ryzen_smu` kernel module to function.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/looking-for-documentation-on-maketargets-within-stdenv-mkderivation/45182/1
